paasta_tools/paasta_maintenance.py

```diff
@@ -0,0 +1,244 @@
+        else:
+            windows = schedule['windows']
+    else:
+        windows = [window]
+
+    payload = dict()
+    payload['windows'] = windows
+
+    return payload
+
+
+def get_maintenance_status(leader):
+    credentials = load_credentials()
+    endpoint = '/master/maintenance/status'
+    url = "http://%s:%s%s" % (leader, PORT, endpoint)
+    return requests.get(url, auth=credentials, timeout=15)
+
+
+def get_maintenance_schedule(leader):
+    credentials = load_credentials()
+    endpoint = '/master/maintenance/schedule'
+    url = "http://%s:%s%s" % (leader, PORT, endpoint)
+    return requests.get(url, auth=credentials, timeout=15)
+
+
+def load_credentials(mesos_secrets='/nail/etc/mesos-secrets'):
+    with open(mesos_secrets) as data_file:
+        data = json.load(data_file)
+    username = data['credentials'][0]['principal']
+    password = data['credentials'][0]['secret']
+    return username, password
+
+
+def send_payload(url, payload):
+    credentials = load_credentials()
+    return requests.post(url, data=json.dumps(payload), auth=credentials, timeout=15)
+
+
+def drain(leader, hostnames, start, duration):
+    payload = build_maintenance_schedule_payload(leader, hostnames, start, duration, drain=True)
+    endpoint = '/master/maintenance/schedule'
+    url = "http://%s:%s%s" % (leader, PORT, endpoint)
+    print send_payload(url, payload).text
+
+
+def undrain(leader, hostnames, start, duration):
+    payload = build_maintenance_schedule_payload(leader, hostnames, start, duration, drain=False)
+    endpoint = '/master/maintenance/schedule'
+    url = "http://%s:%s%s" % (leader, PORT, endpoint)
+    print send_payload(url, payload).text
+
+
+def down(leader, hostnames):
+    payload = build_start_maintenance_payload(hostnames)
+    endpoint = '/master/machine/down'
+    url = "http://%s:%s%s" % (leader, PORT, endpoint)
+    print send_payload(url, payload).text
+
+
+def up(leader, hostnames):
+    payload = build_start_maintenance_payload(hostnames)
+    endpoint = '/master/machine/up'
+    url = "http://%s:%s%s" % (leader, PORT, endpoint)
+    print send_payload(url, payload).text
+
+
+def status(leader):
+    status = get_maintenance_status(leader)
+    print "%s:%s" % (status, status.text)
+
+
+def schedule(leader):
+    schedule = get_maintenance_schedule(leader)
+    print "%s:%s" % (schedule, schedule.text)
+
+
+def paasta_maintenance():
+    """Manipulate the maintenance state of a PaaSTA host."""
+    args = parse_args()
+
+    action = args.action
+    hostnames = args.hostname
+
+    if action not in ['drain', 'undrain', 'down', 'up', 'status']:
+        print "action must be 'drain', 'undrain', 'down', 'up', or 'status'"
+        return
+
+    if action != 'status' and not hostnames:
+        print "You must specify one or more hostnames"
+        return
+
+    try:
+        leader = get_mesos_leader()
+    except MesosMasterConnectionError:
+        print "Run this script from a mesos master!"
```

Comment:
Part of the reason for this is that you can mark a host as being up/down without the box itself being available. i.e. when imaging some new PaaSTA slaves, we've been marking them as down so that they do not immediately register when they come up (we want to wait until Puppet is done running). If we had to SSH to the slave to do this, it would not be possible.